File: leetcode/Python/merge_int.py
# -*- coding: utf-8 -*-
"""
Created on Sat Aug 11 12:02:18 2018
merge intervals
@author: Shubham
"""

import numpy as np
import operator
import functools

class Interval:
    def __init__(self, s = 0, e = 0):
        self.start = s
        self.end = e
           
class Solution:
    def merge(self, intervals):
        ans = []
        sorted(intervals, key=lambda x: x.start)
        start = intervals[0].start
        end = intervals[0].end
        for i in range(1, len(intervals)):
            if end > intervals[i].start:
                end = max(end, intervals[i].end)
            else:
                ans.append(Interval(start, end))
                start = intervals[i].start
                end = intervals[i].end
        ans.append(Intervals(s, e))
        return ans

File: leetcode/C++/merge_int.cpp
#include <iostream>
#include <stdio.h>
using namespace std;

struct Interval
{
	int start, end;
};

bool compareIntervals(const Interval &e1, const Interval &e2)
{
	return e1.start < e2.start;
}

class Solution
{
public:
	vector<Interval> merge(vector<Interval> &intervals)
	{
		vector<Interval> ans;
		if(intervals.empty())
			return ans;

		sort(intervals.begin(), intervals.end(), compareIntervals);
		int s = intervals[0].start;
		int e = intervals[0].end;
		for(int i = 1; i < intervals.size(); i++)
		{
			if(intervals[i].start <= e)
			{
				e = max(e, intervals[i].end);
			}
			else{
				ans.push_back(Interval(s, e));
				s = intervals[i].start;
				e = intervals[i].end;
			}
		}
		ans.push_back(Interval(s, e));
		return ans;
	}
};